### ISSUE TITLE
[Question] How to log an error when Azure App Configuration reload fails

### ISSUE THREAD

**Issue by AlexTorianyk** [Azure/AppConfiguration#713]
I'm using Azure App Configuration with the dynamic configuration using poll model enabled, as described [here](https://learn.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-aspnet-core?tabs=core6x). My goal is to log an error message when the configuration reloads fails for any reason, like an incorrect key-vault reference. Here's the method which adds Azure App Configuration as the configuration source.

```
public static IConfigurationBuilder AddAppConfiguration(IConfigurationBuilder configurationBuilder, IEnumerable<string> filters)
    {
        if (!configurationBuilder.Sources.Any())
        {
            configurationBuilder.AddDefaultConfigurationSources();
        }
        
        var configuration = configurationBuilder.Build();
        var appConfiguration = configuration.GetSection(AppConfiguration.ConfigurationKey)
            .Get<AppConfiguration>();
        
        if (appConfiguration?.IsEnabled == true)
        {
            if (appConfiguration.IsRedundancyEnabled)
                configurationBuilder.AddAzureAppConfiguration(
                    options =>
                    {
                        ConfigureAppConfiguration(options, appConfiguration, appConfiguration.UrlSecondary, filters, appConfiguration.Label);
                    }, true);

            configurationBuilder.AddAzureAppConfiguration(
                options => { ConfigureAppConfiguration(options, appConfiguration, appConfiguration.Url, filters, appConfiguration.Label); },
                true);
        }

        return configurationBuilder;
    }
 ```

Please note that the optional parameter in AddAzureAppConfiguration is set to true. This means that the application won't crash if the configuration refresh fails and will just keep running with the old settings, I'd like to keep that behaviour.

I've tried to use the ChangeToken to log an error if the configuration hasn't changed like this:

```
configuration.GetReloadToken().RegisterChangeCallback(state =>
{
    var reloadToken = state as IChangeToken;
    if (reloadToken.HasChanged)
    {
        logger.Information("Configuration refresh successful");
    }
    else
    {
        logger.Error("Configuration refresh failed");
    }
            
    RegisterCallback(logger, configuration);
}, configuration.GetReloadToken());
```

But the callback only actually gets called if the configuration did in fact change 😅, at least I didn't manage to get it to fail and log the error.

**Comment by amerjusupovic** [comment]
Hi @AlexTorianyk , there is a way to get these kinds of logs explained [here](https://github.com/Azure/AppConfiguration-DotnetProvider/pull/273#issue-972987151), which includes KeyVaultReferenceExceptions for incorrect KeyVault references. Hope this helps!

**Comment by avanigupta** [comment]
In addition to the logging instructions provided above, I wanted to address this point -
> Please note that the optional parameter in AddAzureAppConfiguration is set to true. This means that the application won't crash if the configuration refresh fails and will just keep running with the old settings, I'd like to keep that behaviour.

When optional is set to true, we ensure that the application will not crash if the configuration fails to load during **_App Startup_**. Once the app is up and running, optional flag is ignored. So if there's an error during refresh, your app should keep running with old settings regardless of the optional flag.

**Comment by AlexTorianyk** [comment]
Thanks a lot, @amerjusupovic & @avanigupta, it seems like we have set the Microsoft minimum logging level to _Error_, while the refresh logs were _Warnings_, so they were ignored 😅 

While testing out the logging, I have run into a somewhat related problem though. It seems like the key vault URI that the setting is pointing to doesn't seem to update. I've tried to change _https://bla.vault.azure.net/secrets/StorageAccountKey-valid_ to _https://bla.vault.azure.net/secrets/StorageAccountKey-invalid_ during runtime in order to test whether the refresh fails would log, but it seems like the URI was cached. I've tried to look and debug through the `AzureAppConfigurationProvider`, but the `adapter.InvalidateCache()` is not hit for me. Let me know if you'd like me to create a separate issue for this.

I'd like also to give one more thanks to @avanigupta for clarifying what the optional flag actually does 😊
